src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTip.cs

```diff
@@ -73,10 +129,600 @@ private void AttachToPopup()
+				//    verticalOffset = DEFAULT_MOUSE_OFFSET;
+				//    break;
+				//case AutomaticToolTipInputMode::Touch:
+				//    verticalOffset = DEFAULT_TOUCH_OFFSET;
+				//    break;
+				//}
+
+				verticalOffset = DEFAULT_MOUSE_OFFSET;
+			}
+
+			// UNO TODO
+			//// Reverse Left/Right palacement for RTL, because the target is flipped.
+			//if (bIsRTL)
+			//{
+			//    if (placement == PlacementMode.Right)
+			//    {
+			//        placement = PlacementMode.Left;
+			//    }
+			//    else if (placement == PlacementMode.Left)
+			//    {
+			//        placement = PlacementMode.Right;
+			//    }
+			//}
+
+			// To honor horizontal/vertical offset, inflate the placement target by these offsets before calling into
+			// the Windows popup positioning logic.
+			rcDockTo.Inflate(horizontalOffset, verticalOffset);
+			(var rcResult, var placementChosen) = ToolTipPositioning.QueryRelativePosition(
+				constraint,
+				szFlyout,
+				rcDockTo,
+				placement);
+
+			// if right/bottom doesn't fit into the plug-in bounds, clip the ToolTip
+			{
+				var clipCalculationsRect = new Rect(0, 0, ActualWidth, ActualHeight);
+				CalculateTooltipClip(clipCalculationsRect, visibleRect.Width - visibleRect.X, visibleRect.Height - visibleRect.Y);
+			}
+
+			Console.WriteLine($"wjkreujk rcResult={rcResult}");
+			// Position tooltip by setting popup's offsets
+			spPopup.VerticalOffset = rcResult.Top - origin.Y;
+			// ToolTipPositioning::QueryRelativePosition is used to position in LTR and RTL. In LTR, the horizontal
+			// offset is  in rcResult.Left. In RTL, the horizontal offset is in rcResult.Right because the
+			// popup is flipped.
+			spPopup.HorizontalOffset = (bIsRTL) ? rcResult.Right - origin.X : rcResult.Left - origin.X;
+
+			// There used to be a setting of FromVerticalOffset and FromHorizontalOffset on the ToolTipTemplateSettings
+			// gotten from get_TemplateSettings here.  However, this is no longer done because the ToolTip animation
+			// is now a FadeIn/FadeOut which doesn't use any FromHorizontalOffset/FromVerticalOffset.  This leaves
+			// ToolTipTemplateSettings and all associated code in a basically non-used and deprecated state - but as of
+			// now, we can't make any breaking changes.  So, to preserve max compatibility, the ToolTipTemplateSettings
+			// is still around as a class/interface/property of the tooltip for now.  It should be removed (or at least
+			// the ToolTip-specific-ness of the Tooltip's template settings should be removed) as soon as it's ok to
+			// make a breaking change.
+		}
+
+
+		private void CalculateTooltipClip(
+			Rect toolTipRect, double maxX, double maxY)
+		{
+			var clipSize = default(Size);
+			var dX = 0.0;
+			var dY = 0.0;
+
+			dX = toolTipRect.Left + toolTipRect.Right - maxX;
+			dY = toolTipRect.Top + toolTipRect.Bottom - maxY;
+
+			if ((dX >= 0) || (dY >= 0))
+			{
+				dX = Math.Max(0, dX);
+				dY = Math.Max(0, dY);
+				clipSize.Width = Math.Max(0, toolTipRect.Right - dX);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['clipSize' is null on at least one execution path.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5404644)

Reply:
struct